File: scripts/fetch-latest-spot.js
require('dotenv').config();

const fs = require('fs');
const path = require('path');
const { createCanvas } = require('@napi-rs/canvas');

// Polyfill fetch for Node.js < 18
let fetch;
if (typeof global.fetch === 'undefined') {
    const nodeFetch = require('node-fetch');
    fetch = nodeFetch.default || nodeFetch;
} else {
    fetch = global.fetch;
}

function requireEnv(name) {
    const value = process.env[name];
    if (!value) {
        throw new Error(`Missing environment variable: ${name}`);
    }
    return value;
}

function getEnvOrDefault(name, fallback) {
    return process.env[name] || fallback;
}

function toNumber(value) {
    const n = Number(value);
    if (!Number.isFinite(n)) {
        throw new Error(`Invalid numeric value: ${value}`);
    }
    return n;
}

function normalizeTwelveBar(raw) {
    const timestamp = new Date(`${raw.datetime.replace(' ', 'T')}Z`);
    return {
        timestamp: timestamp.toISOString(),
        minute: new Date(Math.floor(timestamp.getTime() / 60000) * 60000).toISOString(),
        open: toNumber(raw.open),
        high: toNumber(raw.high),
        low: toNumber(raw.low),
        close: toNumber(raw.close),
        volume: 0,
    };
}

async function fetchYesterdaySpotBars() {
    const token = requireEnv('TWELVE_TOKEN');
    const symbol = getEnvOrDefault('TWELVE_SYMBOL', 'XAU/USD');
    const interval = getEnvOrDefault('TWELVE_INTERVAL', '1min');

    // Calculate yesterday's date
    const now = new Date();
    const yesterday = new Date(now);
    yesterday.setDate(yesterday.getDate() - 1);
    
    const year = yesterday.getUTCFullYear();
    const month = String(yesterday.getUTCMonth() + 1).padStart(2, '0');
    const day = String(yesterday.getUTCDate()).padStart(2, '0');
    
    const startDate = `${year}-${month}-${day} 00:00:00`;
    const endDate = `${year}-${month}-${day} 23:59:59`;

    console.log('Fetching yesterday spot data from TwelveData...');
    console.log(`  Symbol: ${symbol}`);
    console.log(`  Interval: ${interval}`);
    console.log(`  Date: ${year}-${month}-${day} (yesterday)`);
    console.log(`  Range: ${startDate} to ${endDate}`);

    const url = new URL('https://api.twelvedata.com/time_series');
    url.searchParams.set('symbol', symbol);
    url.searchParams.set('interval', interval);
    url.searchParams.set('outputsize', '1440'); // Full day: 24 * 60 = 1440 minutes
    url.searchParams.set('apikey', token);
    url.searchParams.set('start_date', startDate);
    url.searchParams.set('end_date', endDate);

    const response = await fetch(url.toString(), {
        headers: {
            'User-Agent': 'gold-tracker-spot-yesterday/1.0',
        },
    });

    if (!response.ok) {
        const text = await response.text();
        throw new Error(`Twelve Data request failed (${response.status}): ${text}`);
    }

    const payload = await response.json();

    if (!payload || payload.status !== 'ok') {
        throw new Error(`Twelve Data request failed: ${JSON.stringify(payload)}`);
    }

    if (!Array.isArray(payload.values) || payload.values.length === 0) {
        throw new Error('No spot data returned for yesterday');
    }

    // Normalize and sort by time (oldest first)
    const bars = payload.values.map(normalizeTwelveBar).sort((a, b) => 
        new Date(a.timestamp) - new Date(b.timestamp)
    );

    console.log(`  Retrieved ${bars.length} bars`);
    console.log(`  Time range: ${bars[0].timestamp} to ${bars[bars.length - 1].timestamp}`);
    
    return bars;
}

function formatTimestamp(value) {
    const date = new Date(value);
    return date.toISOString().replace('T', ' ').replace('Z', ' UTC');
}

function priceToY(price, minPrice, maxPrice, top, height) {
    const span = maxPrice - minPrice;
    if (span === 0) {
        return top + height / 2;
    }
    return top + height - ((price - minPrice) / span) * height;
}

function drawSpotChart(spotBars) {
    const width = 1400;
    const height = 700;
    const canvas = createCanvas(width, height);
    const ctx = canvas.getContext('2d');

    const padLeft = 96;
    const padRight = 24;
    const padTop = 46;
    const padBottom = 80;
    const chartHeight = height - padTop - padBottom;
    const chartWidth = width - padLeft - padRight;

    // Background
    ctx.fillStyle = '#0f172a';
    ctx.fillRect(0, 0, width, height);

    // Title
    ctx.fillStyle = '#f8fafc';
    ctx.font = 'bold 26px sans-serif';
    ctx.fillText('Gold Spot (XAU/USD) Yesterday Full-Day K-Line', padLeft, 30);

    // Subtitle
    ctx.font = '14px sans-serif';
    ctx.fillStyle = '#cbd5e1';
    ctx.fillText(`${spotBars.length} bars | ${formatTimestamp(spotBars[0].timestamp)} -> ${formatTimestamp(spotBars[spotBars.length - 1].timestamp)}`, padLeft, 62);

    // Calculate price range
    const allPrices = spotBars.flatMap((bar) => [bar.open, bar.high, bar.low, bar.close]);
    const minPrice = Math.min(...allPrices);
    const maxPrice = Math.max(...allPrices);
    const span = Math.max(1, maxPrice - minPrice);
    const padding = span * 0.05;
    const plotMin = minPrice - padding;
    const plotMax = maxPrice + padding;

    // Draw grid lines
    ctx.strokeStyle = '#334155';
    ctx.lineWidth = 1;
    for (let i = 0; i <= 5; i += 1) {
        const y = padTop + (chartHeight / 5) * i;
        ctx.beginPath();
        ctx.moveTo(padLeft, y);
        ctx.lineTo(width - padRight, y);
        ctx.stroke();
        
        ctx.fillStyle = '#cbd5e1';
        ctx.font = '13px sans-serif';
        ctx.fillText((plotMax - ((plotMax - plotMin) / 5) * i).toFixed(2), 12, y + 4);
    }

    const candleWidth = Math.max(2, Math.floor(chartWidth / Math.max(spotBars.length, 1) * 0.55));

    // Draw candlesticks
    for (let i = 0; i < spotBars.length; i += 1) {
        const bar = spotBars[i];
        const x = padLeft + i * (chartWidth / Math.max(spotBars.length - 1, 1));
        const openY = priceToY(bar.open, plotMin, plotMax, padTop, chartHeight);
        const closeY = priceToY(bar.close, plotMin, plotMax, padTop, chartHeight);
        const highY = priceToY(bar.high, plotMin, plotMax, padTop, chartHeight);
        const lowY = priceToY(bar.low, plotMin, plotMax, padTop, chartHeight);

        // Wick
        ctx.strokeStyle = '#38bdf8';
        ctx.lineWidth = 1;
        ctx.beginPath();
        ctx.moveTo(x, highY);
        ctx.lineTo(x, lowY);
        ctx.stroke();

        // Body
        ctx.fillStyle = '#38bdf8';
        ctx.fillRect(x - candleWidth / 2, Math.min(openY, closeY), candleWidth, Math.max(2, Math.abs(closeY - openY)));

        // Time labels
        if (i % Math.max(1, Math.floor(spotBars.length / 10)) === 0) {
            ctx.fillStyle = '#cbd5e1';
            ctx.font = '12px sans-serif';
            ctx.textAlign = 'center';
            ctx.fillText(formatTimestamp(bar.timestamp).slice(5, 16), x, height - 20);
        }
    }

    // Legend
    ctx.fillStyle = '#cbd5e1';
    ctx.font = '13px sans-serif';
    ctx.fillText('Spot (XAU/USD)', padLeft + 120, 100);
    ctx.fillStyle = '#38bdf8';
    ctx.fillRect(padLeft + 70, 88, 12, 12);

    return canvas;
}

async function main() {
    try {
        // Step 1: Fetch yesterday spot data
        const spotBars = await fetchYesterdaySpotBars();

        // Step 2: Save to CSV
        console.log('\nSaving spot data to CSV...');
        const csvRows = ['timestamp,minute,open,high,low,close,volume'];
        for (const bar of spotBars) {
            csvRows.push(`${bar.timestamp},${bar.minute},${bar.open},${bar.high},${bar.low},${bar.close},${bar.volume}`);
        }
        const csvPath = getEnvOrDefault('YESTERDAY_SPOT_CSV_PATH', path.join(__dirname, '..', 'yesterday-spot-data.csv'));
        fs.writeFileSync(csvPath, csvRows.join('\n'), 'utf8');
        console.log(`Saved to: ${csvPath}`);

        // Step 3: Draw chart
        console.log('\nDrawing spot chart...');
        const canvas = drawSpotChart(spotBars);
        const outputPath = getEnvOrDefault('YESTERDAY_SPOT_PLOT_PATH', path.join(__dirname, '..', 'yesterday-spot-kline.png'));
        fs.writeFileSync(outputPath, canvas.toBuffer('image/png'));
        console.log(`Chart saved to: ${outputPath}`);

        // Summary
        console.log('\n=== Summary ===');
        console.log(`Total bars: ${spotBars.length}`);
        console.log(`Price range: $${Math.min(...spotBars.map(b => b.close)).toFixed(2)} - $${Math.max(...spotBars.map(b => b.close)).toFixed(2)}`);
        console.log(`Time range: ${spotBars[0].timestamp.split('T')[1].slice(0, 5)} - ${spotBars[spotBars.length - 1].timestamp.split('T')[1].slice(0, 5)} UTC`);
        console.log('\n✓ Latest spot data processing complete!');

    } catch (error) {
        console.error('Error:', error.message);
        process.exitCode = 1;
    }
}

main();

File: scripts/FetchLatestSpot.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace GoldTracker
{
    public class SpotBar
    {
        public string Timestamp { get; set; }
        public string Minute { get; set; }
        public double Open { get; set; }
        public double High { get; set; }
        public double Low { get; set; }
        public double Close { get; set; }
        public double Volume { get; set; }
    }

    public class FetchLatestSpot
    {
        private static string RequireEnv(string name)
        {
            var value = Environment.GetEnvironmentVariable(name);
            if (string.IsNullOrEmpty(value))
            {
                throw new Exception("Missing environment variable: " + name);
            }
            return value;
        }

        private static string GetEnvOrDefault(string name, string fallback)
        {
            var value = Environment.GetEnvironmentVariable(name);
            return string.IsNullOrEmpty(value) ? fallback : value;
        }

        private static double ToNumber(object value)
        {
            double result;
            if (!double.TryParse(value.ToString(), out result))
            {
                throw new Exception("Invalid numeric value: " + value);
            }
            return result;
        }

        private static SpotBar NormalizeTwelveBar(JObject raw)
        {
            var datetime = raw["datetime"].ToString();
            var timestamp = DateTime.Parse(datetime.Replace(" ", "T") + "Z");
            
            var minute = new DateTime(
                timestamp.Year, 
                timestamp.Month, 
                timestamp.Day, 
                timestamp.Hour, 
                timestamp.Minute, 
                0, 
                DateTimeKind.Utc
            );

            return new SpotBar
            {
                Timestamp = timestamp.ToString("yyyy-MM-ddTHH:mm:ssZ"),
                Minute = minute.ToString("yyyy-MM-ddTHH:mm:ssZ"),
                Open = ToNumber(raw["open"]),
                High = ToNumber(raw["high"]),
                Low = ToNumber(raw["low"]),
                Close = ToNumber(raw["close"]),
                Volume = 0
            };
        }

        private static List<SpotBar> FetchYesterdaySpotBars()
        {
            var token = RequireEnv("TWELVE_TOKEN");
            var symbol = GetEnvOrDefault("TWELVE_SYMBOL", "XAU/USD");
            var interval = GetEnvOrDefault("TWELVE_INTERVAL", "1min");

            // Calculate yesterday's date
            var now = DateTime.UtcNow;
            var yesterday = now.AddDays(-1);
            var dateStr = yesterday.ToString("yyyy-MM-dd");
            var startDate = dateStr + " 00:00:00";
            var endDate = dateStr + " 23:59:59";

            Console.WriteLine("Fetching yesterday spot data from TwelveData...");
            Console.WriteLine("  Symbol: " + symbol);
            Console.WriteLine("  Interval: " + interval);
            Console.WriteLine("  Date: " + dateStr + " (yesterday)");
            Console.WriteLine("  Range: " + startDate + " to " + endDate);

            var url = "https://api.twelvedata.com/time_series" +
                "?symbol=" + Uri.EscapeDataString(symbol) +
                "&interval=" + Uri.EscapeDataString(interval) +
                "&outputsize=1440" +
                "&apikey=" + Uri.EscapeDataString(token) +
                "&start_date=" + Uri.EscapeDataString(startDate) +
                "&end_date=" + Uri.EscapeDataString(endDate);

            var request = WebRequest.Create(url) as HttpWebRequest;
            request.UserAgent = "gold-tracker-spot-yesterday/1.0";
            request.Method = "GET";

            string responseText;
            using (var response = request.GetResponse())
            using (var reader = new StreamReader(response.GetResponseStream()))
            {
                responseText = reader.ReadToEnd();
            }

            var payload = JsonConvert.DeserializeObject<JObject>(responseText);

            if (payload == null || payload["status"].ToString() != "ok")
            {
                throw new Exception("Twelve Data request failed: " + responseText);
            }

            var values = payload["values"] as JArray;
            if (values == null || values.Count == 0)
            {
                throw new Exception("No spot data returned for yesterday");
            }

            var bars = new List<SpotBar>();
            foreach (JObject item in values)
            {
                bars.Add(NormalizeTwelveBar(item));
            }

            bars.Sort((a, b) => DateTime.Compare(
                DateTime.Parse(a.Timestamp),
                DateTime.Parse(b.Timestamp)
            ));

            Console.WriteLine("  Retrieved " + bars.Count + " bars");
            Console.WriteLine("  Time range: " + bars[0].Timestamp + " to " + bars[bars.Count - 1].Timestamp);

            return bars;
        }

        private static string FormatTimestamp(string value)
        {
            var date = DateTime.Parse(value);
            return date.ToString("yyyy-MM-dd HH:mm:ss") + " UTC";
        }

        private static double PriceToY(double price, double minPrice, double maxPrice, double top, double height)
        {
            var span = maxPrice - minPrice;
            if (span == 0)
            {
                return top + height / 2;
            }
            return top + height - ((price - minPrice) / span) * height;
        }

        private static void DrawSpotChart(List<SpotBar> spotBars, string outputPath)
        {
            var width = 1400;
            var height = 700;
            
            // Create bitmap (similar to canvas)
            var bitmap = new Bitmap(width, height);
            var g = Graphics.FromImage(bitmap);

            var padLeft = 96;
            var padRight = 24;
            var padTop = 46;
            var padBottom = 80;
            var chartHeight = height - padTop - padBottom;
            var chartWidth = width - padLeft - padRight;

            // Background
            g.FillRectangle(new SolidBrush(Color.FromArgb(15, 23, 42)), 0, 0, width, height);

            // Title
            g.DrawString("Gold Spot (XAU/USD) Yesterday Full-Day K-Line",
                new Font("Arial", 26, FontStyle.Bold),
                new SolidBrush(Color.FromArgb(248, 250, 252)),
                padLeft, 30);

            // Subtitle
            g.DrawString(spotBars.Count + " bars | " +
                FormatTimestamp(spotBars[0].Timestamp) + " -> " +
                FormatTimestamp(spotBars[spotBars.Count - 1].Timestamp),
                new Font("Arial", 14),
                new SolidBrush(Color.FromArgb(203, 213, 225)),
                padLeft, 62);

            // Calculate price range
            var allPrices = new List<double>();
            foreach (var bar in spotBars)
            {
                allPrices.Add(bar.Open);
                allPrices.Add(bar.High);
                allPrices.Add(bar.Low);
                allPrices.Add(bar.Close);
            }
            var minPrice = allPrices.Min();
            var maxPrice = allPrices.Max();
            var span = Math.Max(1, maxPrice - minPrice);
            var padding = span * 0.05;
            var plotMin = minPrice - padding;
            var plotMax = maxPrice + padding;

            // Draw grid lines
            for (var i = 0; i <= 5; i++)
            {
                var y = padTop + (chartHeight / 5.0) * i;
                g.DrawLine(new Pen(Color.FromArgb(51, 65, 85)), padLeft, y, width - padRight, y);
                
                var priceLabel = (plotMax - ((plotMax - plotMin) / 5.0) * i).ToString("F2");
                g.DrawString(priceLabel,
                    new Font("Arial", 13),
                    new SolidBrush(Color.FromArgb(203, 213, 225)),
                    12, (float)y + 4);
            }

            var candleWidth = Math.Max(2, (int)Math.Floor(chartWidth / (double)Math.Max(spotBars.Count, 1) * 0.55));

            // Draw candlesticks
            for (var i = 0; i < spotBars.Count; i++)
            {
                var bar = spotBars[i];
                var x = padLeft + i * (chartWidth / (double)Math.Max(spotBars.Count - 1, 1));
                var openY = PriceToY(bar.Open, plotMin, plotMax, padTop, chartHeight);
                var closeY = PriceToY(bar.Close, plotMin, plotMax, padTop, chartHeight);
                var highY = PriceToY(bar.High, plotMin, plotMax, padTop, chartHeight);
                var lowY = PriceToY(bar.Low, plotMin, plotMax, padTop, chartHeight);

                // Wick
                g.DrawLine(new Pen(Color.FromArgb(56, 189, 248)),
                    (float)x, (float)highY, (float)x, (float)lowY);

                // Body
                var bodyTop = Math.Min(openY, closeY);
                var bodyHeight = Math.Max(2, Math.Abs(closeY - openY));
                g.FillRectangle(new SolidBrush(Color.FromArgb(56, 189, 248)),
                    (float)(x - candleWidth / 2.0), (float)bodyTop,
                    candleWidth, (float)bodyHeight);

                // Time labels
                if (i % Math.Max(1, (int)Math.Floor(spotBars.Count / 10.0)) == 0)
                {
                    var timeStr = FormatTimestamp(bar.Timestamp).Substring(5, 11);
                    var format = new StringFormat { Alignment = StringAlignment.Center };
                    g.DrawString(timeStr,
                        new Font("Arial", 12),
                        new SolidBrush(Color.FromArgb(203, 213, 225)),
                        (float)x, height - 20, format);
                }
            }

            // Legend
            g.DrawString("Spot (XAU/USD)",
                new Font("Arial", 13),
                new SolidBrush(Color.FromArgb(203, 213, 225)),
                padLeft + 120, 100);
            g.FillRectangle(new SolidBrush(Color.FromArgb(56, 189, 248)),
                padLeft + 70, 88, 12, 12);

            // Save to file
            bitmap.Save(outputPath, ImageFormat.Png);
            bitmap.Dispose();
            g.Dispose();
        }

        public static void Main(string[] args)
        {
            try
            {
                // Step 1: Fetch yesterday spot data
                var spotBars = FetchYesterdaySpotBars();

                // Step 2: Save to CSV
                Console.WriteLine("\nSaving spot data to CSV...");
                var csvRows = new List<string>();
                csvRows.Add("timestamp,minute,open,high,low,close,volume");
                
                foreach (var bar in spotBars)
                {
                    csvRows.Add(string.Format("{0},{1},{2},{3},{4},{5},{6}",
                        bar.Timestamp, bar.Minute, bar.Open, bar.High, bar.Low, bar.Close, bar.Volume));
                }

                var csvPath = GetEnvOrDefault("YESTERDAY_SPOT_CSV_PATH", 
                    Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "output", "csv", "yesterday-spot-data.csv"));
                File.WriteAllLines(csvPath, csvRows.ToArray(), Encoding.UTF8);
                Console.WriteLine("Saved to: " + csvPath);

                // Step 3: Draw chart (placeholder)
                Console.WriteLine("\nDrawing spot chart...");
                var outputPath = GetEnvOrDefault("YESTERDAY_SPOT_PLOT_PATH",
                    Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "output", "yesterday-spot-kline.png"));
                DrawSpotChart(spotBars, outputPath);

                // Summary
                Console.WriteLine("\n=== Summary ===");
                Console.WriteLine("Total bars: " + spotBars.Count);
                
                var closePrices = spotBars.Select(b => b.Close).ToList();
                Console.WriteLine("Price range: $" + closePrices.Min().ToString("F2") + 
                    " - $" + closePrices.Max().ToString("F2"));
                
                var firstTime = spotBars[0].Timestamp.Split('T')[1].Substring(0, 5);
                var lastTime = spotBars[spotBars.Count - 1].Timestamp.Split('T')[1].Substring(0, 5);
                Console.WriteLine("Time range: " + firstTime + " - " + lastTime + " UTC");
                Console.WriteLine("\n✓ Latest spot data processing complete!");
            }
            catch (Exception error)
            {
                Console.Error.WriteLine("Error: " + error.Message);
                Environment.ExitCode = 1;
            }
        }
    }
}
